### PR TITLE
Adds an Action page

### PR DESCRIPTION
**Pull request by Tazaf**
Adds a new directive to manage broken image link and replace them with a
default image, stored locally.
Also added a new typescript alias to the new `directives` folder.

Integrates the mockup for the Action page.

Adds a new page for the Theme, but it's currently blank. This will be
implemented in the following PR.

Adds some native plugins to the app, that are not necessarily used now
(like the Toast one) but will be in a not so distant future.

Implements some features regarding the Status Bar, that will could be
enhanced in future PR.

Updates unit and e2e test to include the new Action page.
Also makes small updates on previously existing tests.

STORY : TG-55

### CODE REVIEW

File: config.xml [hunk 1]
```diff
@@ -78,6 +79,7 @@
         <splash height="480" src="resources/ios/splash/Default~iphone.png" width="320" />
         <splash height="2732" src="resources/ios/splash/Default@2x~universal~anyany.png" width="2732" />
     </platform>
+    <allow-navigation href="http://10.192.114.209:8100" sessionid="5758039e" />
```
Comment (AlphaHydrae) [inline config.xml]: Is this necessary?

File: src/pages/action/action.html [hunk 2]
```diff
@@ -1,12 +1,35 @@
...
-
-</ion-content>
+  <header>
+    <div class="img-wrapper">
+      <img [src]="action.photoUrl" [alt]="action.title" default="assets/img/default_action.jpg" class="action-cover">
+    </div>
+    <section class="title-bar">
+      <div>
+        <h1 class="action-title">{{ action.title }}</h1>
+        <h2 class="theme-title" [navPush]="themePage" [navParams]="{ theme: action.theme }">{{ action.theme.title }}
```
Comment (AlphaHydrae) [inline src/pages/action/action.html]: I'm highlighting this here but there's also a similar page transition in the previous PR.

We now have two ways to make page transitions in the code:

1. Inject `NavController` and use its `push` method in a component's TypeScript code (used in the map page to transition to the actions list).
2. Use `[navPush]` in a component's HTML template (used here and in the previous PR).

Unless there's a good reason to use different techniques, we should stick to one.

My gut feeling is that I prefer solution 1 with a `goToThatPage` method like we've got in the map page. Solution 2 forces us to add an attribute just to pass the target page to the view, which feels dirty to me. What do you think?

File: e2e/app.spec.ts [hunk 3]
```diff
@@ -142,5 +147,22 @@ describe('App', function() {
     // Ensure that more actions have been loaded on the page.
     actionListItemsFinder = await actionsListPage.getActionListItems();
     expect(actionListItemsFinder).to.have.lengthOf(6);
+
+    // Click on the first action item
+    actionListItemsFinder[0].click();
```
Comment (AlphaHydrae) [inline e2e/app.spec.ts]: My understanding of the page object pattern is that we should avoid doing this kind of thing (clicking, interacting with the DOM) in the test itself, but rather add a `goToAction(n)` method in the `ActionsListPageObject ` class to do it for us, which (hopefully) decouples the test itself from the DOM (at least a little more).

This `goToAction(n)` method would retrieve the list, click on the n-th action and directly return the `ActionPageObject` representing the new page, basically this:

```
async goToAction(n) {

  const actions = await actionsListPage.getActionListItems();
  actionListItemsFinder[0].click();

  const actionPage = new ActionPageObject('action-page');
  await presenceOf(actionPage.getPage());

  return actionPage;
}
```

Note that I have no problem with what the test does, it's just a pattern/style concern.

File: e2e/app.spec.ts [hunk 4]
```diff
@@ -34,20 +35,23 @@ const ONEX_BBOX_HEIGHT = distance(point(ONEX_BBOX.southWest), point([ ONEX_BBOX.
...
-  let locations;
+  let locations, actions;
   let mapPage: MapPageObject;
   let actionsListPage: ActionsListPageObject;
+  let actionPage: ActionPageObject;
 
   beforeEach(async function() {
     mapPage = new MapPageObject('map-page');
     actionsListPage = new ActionsListPageObject('actions-list-page');
+    actionPage = new ActionPageObject('action-page');
```
Comment (AlphaHydrae) [inline e2e/app.spec.ts]: I'm only noticing this now, and I may well have started it, but I don't think we should have to specify the selector here in each test. It should be a default value in those classes' constructors.

File: src/pages/theme/theme.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+export class ThemePage {
+
+  constructor(
+    public navCtrl: NavController,
+    public navParams: NavParams
+  ) { }
+
+  ionViewDidLoad() {
+    // tslint:disable-next-line:no-console
+    console.log('ionViewDidLoad ThemePage', this.navParams.get('theme'));
```
Comment (AlphaHydrae) [inline src/pages/theme/theme.ts]: Leftover log not using Print.

File: src/providers/actions-service/actions-service.ts [hunk 6]
```diff
@@ -13,7 +13,7 @@ interface FetchActionsParams {
   offset?: string;
   limit?: string;
   include?: string | string[];
-  [key: string]: string | string[];
+  [ key: string ]: string | string[];
```
Comment (AlphaHydrae) [inline src/providers/actions-service/actions-service.ts]:
```suggestion
  [key: string]: string | string[];
```

This is an index signature, not an array.

File: src/providers/actions-service/actions-service.ts [hunk 7]
```diff
@@ -33,9 +32,17 @@ export default class ActionsService {
...
+      .pipe(map(response => new PaginatedResponse<Action>(response, parseApiAction)));
+  }
+
+  /**
+   * Fetchs one action, through its ID, from the BioPocket API.
+   * @param actionId The id of the action to fetch.
+   */
+  fetchAction(actionId: string): Observable<Action> {
+    return this.http
+      .get<Action>(`${RESOURCE_PATH}/${actionId}`)
```
Comment (AlphaHydrae) [inline src/providers/actions-service/actions-service.ts]:
```suggestion
      .get(`${RESOURCE_PATH}/${actionId}`)
```

It's not an `Action` until it's passed through `parseApiAction`.

**Review by AlphaHydrae — CHANGES_REQUESTED** [review]

File: config.xml (hunk 1 above)
Comment (Tazaf) [reply config.xml]: Dunno. Might be a leftover from a live-reload configuration made by Ionic. Will remove.

Comment (Tazaf) [reply config.xml]: Dunno. Might be a leftover from a live-reload configuration made by Ionic. Will remove.

File: e2e/app.spec.ts (hunk 4 above)
Comment (Tazaf) [reply e2e/app.spec.ts]: It is now ;)
I've come to the same conclusion in the last couple of PR, and made the change there (among other changes)

File: src/pages/theme/theme.ts (hunk 5 above)
Comment (Tazaf) [reply src/pages/theme/theme.ts]: Will remove the method.

File: e2e/app.spec.ts (hunk 3 above)
Comment (Tazaf) [reply e2e/app.spec.ts]: Okay. I tried implementing something like this in the latest commits. Tell me if this is what you had in mind.
Maybe we ought to extend this pattern to other interactions done in the test ? (like clicking on the map markers or the popover backdrop) ?

Comment (Tazaf) [reply e2e/app.spec.ts]: > Maybe we ought to extend this pattern to other interactions done in the test ?

Well, you responded to that in your comments on the next PR. I will try something on that one, then.

**Pull request merged**